core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,425 @@
+    isDirectBuf: Boolean)
+    extends FilterInputStream(in: InputStream) with ReadableByteChannel {
+  val oneByteBuf = new Array[Byte](1)
+
+  val bufferSize = CryptoStreamUtils.checkBufferSize(codec, bufferSizeVal)
+  /**
+   * Input data buffer. The data starts at inBuffer.position() and ends at
+   * to inBuffer.limit().
+   */
+  val inBuffer = if (isDirectBuf) {
+    ByteBuffer.allocateDirect(bufferSizeVal)
+  } else {
+    ByteBuffer.allocate(bufferSizeVal)
+  }
+
+  /**
+   * The decrypted data buffer. The data starts at outBuffer.position() and
+   * ends at outBuffer.limit()
+   */
+  val outBuffer = if (isDirectBuf) {
+    ByteBuffer.allocateDirect(bufferSizeVal)
+  } else {
+    ByteBuffer.allocate(bufferSizeVal)
+  }
+
+  /**
+   * Whether the underlying stream supports
+   * [[org.apache.hadoop.fs.ByteBufferReadable]]
+   */
+  var usingByteBufferRead = false
+  var usingByteBufferReadInitialized = false
+  /**
+   * Padding = pos%(algorithm blocksize) Padding is put into [[inBuffer]]
+   * before any other data goes in. The purpose of padding is to put the input
+   * data at proper position.
+   */
+  var padding: Byte = '0'
+  var closed: Boolean = false
+  var key: Array[Byte] = keyVal.clone()
+  var initIV: Array[Byte] = ivVal.clone()
+  var iv: Array[Byte] = ivVal.clone()
+  var isReadableByteChannel: Boolean = in.isInstanceOf[ReadableByteChannel]
+
+  /** DirectBuffer pool */
+  var bufferPool: Queue[ByteBuffer] = new ConcurrentLinkedQueue[ByteBuffer]()
+  /** Decryptor pool */
+  var decryptorPool: Queue[Decryptor] = new ConcurrentLinkedQueue[Decryptor]()
+
+  lazy val tmpBuf: Array[Byte] = new Array[Byte](bufferSize)
+  var decryptor: Decryptor = getDecryptor()
+  CryptoStreamUtils.checkCodec(codec)
+  resetStreamOffset(streamOffset)
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      bufferSize: Integer,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, bufferSize, key, iv, 0, true)
+  }
+
+  def this(in: InputStream,
+      codec: CryptoCodec,
+      key: Array[Byte],
+      iv: Array[Byte]) {
+    this(in, codec, CryptoStreamUtils.getBufferSize, key, iv)
+  }
+
+  def getWrappedStream(): InputStream = in
+
+  /**
+   * Decryption is buffer based.
+   * If there is data in [[outBuffer]], then read it out of this buffer.
+   * If there is no data in [[outBuffer]], then read more from the
+   * underlying stream and do the decryption.
+   * @param b the buffer into which the decrypted data is read.
+   * @param off the buffer offset.
+   * @param len the maximum number of decrypted data bytes to read.
+   * @return int the total number of decrypted data bytes read into the buffer.
+   * @throws IOException
+   */
+  override def read(b: Array[Byte], off: Int, len: Int): Int = {
+    checkStream()
+    Preconditions.checkNotNull(b)
+    Preconditions.checkArgument(!(off < 0 || len < 0 || len > b.length - off))
+
+    if (len == 0) {
```

Comment:
This is one case where an early return helps in cleaning up the code by avoiding needless indentation.

    if (len == 0) {
      return 0
    }

    val remaining ....